File: wsb_signals/aggregate.py
"""Windowed aggregator — mentions → (ticker, window) empirical features (Phase 1).

For the current clock-aligned tumbling window it computes sov (PRIMARY rank), velocity/accel,
net_dir, dd_count, flair_counts, a baseline-gated z, and the composite H_e (signal-framework
§2–§5). SoV-primary: z is COMPUTED but kept off the H_e blend until its hour-of-week baseline is
`ready` (config weight stays 0 until then). Forward-only — baselines warm from live data, no
backfill. H_e components are standardized cross-sectionally by max-normalization (scaled by the
window max, negatives floored to 0) so they share a [0,1] scale before the weighted blend — see
`_max_norm` for why max-norm, not percentile rank.
"""
from __future__ import annotations

import json
import time
from collections import Counter, defaultdict
from pathlib import Path

from .db import DB, pretty_name
from .models import EmpiricalFeature


def hour_of_week(epoch: int) -> int:
    """0..167 — Monday 00:00 UTC = 0 (session-aware baselines, signal-framework §4)."""
    t = time.gmtime(epoch)
    return t.tm_wday * 24 + t.tm_hour


def window_start_for(now: int, window_seconds: int) -> int:
    """Clock-aligned tumbling window containing `now`."""
    return (now // window_seconds) * window_seconds


def _max_norm(values: list[float]) -> list[float]:
    """Scale to [0,1] by the window max (negatives floored to 0); all-zero → zeros.

    Max-norm — NOT percentile rank — because H_e must stay SoV-primary: percentile rank
    flattens 1st-vs-2nd SoV into a tiny gap, letting secondaries (net_dir) override the
    primary signal, and it hands tied all-zero components (first-window rank_delta/dd_count)
    a spurious 1.0. Max-norm preserves the leader's magnitude and zeroes empty components.
    """
    vmax = max(values, default=0.0)
    if vmax <= 0:
        return [0.0] * len(values)
    return [max(0.0, v) / vmax for v in values]


def aggregate_window(
    db: DB,
    window_start: int,
    window_seconds: int,
    *,
    weights: dict,
    min_samples_ready: int,
    min_authors_full: float = 1.0,
) -> list[EmpiricalFeature]:
    end = window_start + window_seconds
    rows = db.mentions_in_window(window_start, end)
    if not rows:
        return []

    # group by ticker, dedup things by thing_id (one mention per (ticker, thing) — §2.1 grain)
    per: dict[str, dict] = defaultdict(
        lambda: {"things": set(), "authors": set(), "bull": 0, "bear": 0, "dd": set(), "flairs": Counter()}
    )
    for ticker, thing_id, thing_type, author, flair, direction in rows:
        d = per[ticker]
        if thing_id in d["things"]:
            continue
        d["things"].add(thing_id)
        if author:
            d["authors"].add(author)
        if direction == "bull":
            d["bull"] += 1
        elif direction == "bear":
            d["bear"] += 1
        if flair:
            d["flairs"][flair] += 1
            if thing_type == "post" and flair.strip().upper() == "DD":
                d["dd"].add(thing_id)

    total = sum(len(d["things"]) for d in per.values()) or 1
    prior = db.features_at(window_start - window_seconds)         # mentions(W−1), velocity(W−1)
    prior_exists = bool(prior)                                    # was W−1 aggregated at all?
    prior_ranks = db.sov_ranks_at(window_start - window_seconds)  # rank(W−1)

    # baselines: prior windows in the same hour-of-week bucket (forward-only)
    how = hour_of_week(window_start)
    base: dict[str, list[int]] = defaultdict(list)
    for tk, ws, m in db.feature_history(window_start):
        if hour_of_week(ws) == how:
            base[tk].append(m)

    # first pass — raw features per ticker
    feats: list[dict] = []
    for t, d in per.items():
        m = len(d["things"])
        if prior_exists:
            # W−1 was aggregated, so a ticker absent from it genuinely had 0 mentions then
            # (velocity = m is a real new-arrival breakout). accel needs W−1's stored velocity,
            # which is itself None if W−2 was missing → then accel is undefined.
            pm = prior.get(t, {}).get("mentions") or 0
            velocity = float(m - pm)
            pv = prior.get(t, {}).get("velocity")
            accel = float(velocity - pv) if pv is not None else None
        else:
            # No prior window at all (cold start, or a gap in the run loop). velocity/accel are
            # UNDEFINED: emitting 0-based deltas here would make EVERY ticker look like a fresh
            # breakout (velocity=m, accel=m) and inflate H_e. Leave them null → they contribute
            # nothing to the blend (signal-framework §2.2 — momentum needs a real predecessor).
            velocity = accel = None
        bull, bear = d["bull"], d["bear"]
        net_dir = (bull - bear) / (bull + bear) if (bull + bear) else 0.0
        samples = base.get(t, [])
        status, z = "cold", None
        # max(2, …): the sample variance divides by (n−1), so a misconfigured min_samples_ready=1
        # would be a ZeroDivisionError — never treat a single sample as `ready`.
        if len(samples) >= max(2, min_samples_ready):
            status = "ready"
            mean = sum(samples) / len(samples)
            var = sum((x - mean) ** 2 for x in samples) / (len(samples) - 1)
            sd = var ** 0.5
            z = (m - mean) / sd if sd > 0 else None
        elif samples:
            status = "warming"
        feats.append({
            "ticker": t, "mentions": m, "authors": len(d["authors"]), "sov": m / total,
            "velocity": velocity, "accel": accel, "net_dir": net_dir, "dd_count": len(d["dd"]),
            # sort_keys → canonical JSON: the stored string can't depend on flair-encounter order
            # (which would otherwise vary with DB row order). Determinism, not cosmetics.
            "flair_counts": json.dumps(dict(d["flairs"]), sort_keys=True), "z": z, "baseline_status": status,
        })

    # rank_delta: prior rank − current rank (+ve = climbing the SoV board). Tie-break equal-SoV
    # tickers by `ticker` so the rank (and thus rank_delta) is deterministic, not dependent on
    # mention/row iteration order.
    cur_rank = {f["ticker"]: r for r, f in enumerate(sorted(feats, key=lambda f: (-f["sov"], f["ticker"])), 1)}
    for f in feats:
        pr = prior_ranks.get(f["ticker"])
        f["rank_delta"] = float(pr - cur_rank[f["ticker"]]) if pr else 0.0

    # normalize components to [0,1] by window max, then weighted blend → H_e (SoV-primary)
    sov_n = _max_norm([f["sov"] for f in feats])
    acc_n = _max_norm([f["accel"] if f["accel"] is not None else 0.0 for f in feats])  # None/neg → 0
    rd_n = _max_norm([f["rank_delta"] for f in feats])        # all-zero (first window) → zeros
    au_n = _max_norm([float(f["authors"]) for f in feats])
    dd_n = _max_norm([float(f["dd_count"]) for f in feats])   # no DD posts → zeros
    z_n = _max_norm([f["z"] if f["z"] is not None else 0.0 for f in feats])

    w = weights
    out: list[EmpiricalFeature] = []
    for i, f in enumerate(feats):
        z_ready = f["baseline_status"] == "ready" and f["z"] is not None
        h_e = (
            w["sov"] * sov_n[i]
            + w["accel"] * acc_n[i]
            + w["rank_delta"] * rd_n[i]
            + w["authors"] * au_n[i]
            + w["conviction"] * dd_n[i]
            + w["net_dir"] * abs(f["net_dir"])
            + (w["z"] * z_n[i] if z_ready else 0.0)
        )
        # Support shrink: a thin-evidence row (1–2 distinct authors) must not top the board off
        # max-normalized momentum alone — in a quiet window a single mention max-norms to 1.0 on
        # every component. Damp the *absolute* H_e toward 0 until `min_authors_full` distinct
        # authors back it. This is an absolute-confidence axis, distinct from the relative
        # `authors` component above (signal-framework §4: rank on robust support, not raw counts).
        if min_authors_full > 0:
            h_e *= min(1.0, f["authors"] / min_authors_full)
        out.append(EmpiricalFeature(
            ticker=f["ticker"], window_start=window_start, mentions=f["mentions"],
            authors=f["authors"], sov=f["sov"], velocity=f["velocity"], accel=f["accel"],
            z=f["z"], net_dir=f["net_dir"], dd_count=f["dd_count"],
            flair_counts=f["flair_counts"], baseline_status=f["baseline_status"], h_e=h_e,
        ))
    # Canonical board order: H_e-primary, with an explicit total-order tie-break chain down to
    # `ticker` so equal-H_e rows have ONE deterministic ordering — reproducible run-to-run and
    # well-defined for any future re-implementation (the board must not depend on DB row order or
    # dict-insertion order). Ties cascade H_e → sov → authors → mentions → ticker.
    out.sort(key=lambda r: (-r.h_e, -r.sov, -r.authors, -r.mentions, r.ticker))
    return out


def write_snapshot(
    rows: list[EmpiricalFeature],
    window_start: int,
    window_seconds: int,
    path: Path,
    generated_at: int,
    analytical: dict | None = None,
    movers: list | None = None,
    names: dict | None = None,
    min_window_mentions: int | None = None,
    capped: bool = False,
) -> None:
    """Write the dashboard's JSON contract (the daemon's output; avoids DuckDB write-lock contention).

    `analytical` maps ticker → AnalyticalFeature (Phase-2 market overlay); market fields are null
    for tickers outside the fetched top-N. `movers` is the screener list (for the STEALTH teaser).
    `names` maps symbol → raw company name (from ticker_names); displayed alongside every ticker.
    `min_window_mentions`: when the window's TOTAL mentions fall below this, the snapshot is flagged
    `quiet` so the dashboard can banner the board as low-confidence (off-hours single-mention noise).
    `capped`: the source poll hit its pagination cap (window incomplete) → flagged so the board can
    banner the SoV as untrustworthy (data-model invariant 14).
    """
    analytical = analytical or {}
    names = names or {}
    total_mentions = sum(r.mentions for r in rows)
    out_rows = []
    for i, r in enumerate(rows, 1):
        a = analytical.get(r.ticker)
        out_rows.append({
            "rank": i, "ticker": r.ticker, "name": pretty_name(names.get(r.ticker)),
            "mentions": r.mentions, "authors": r.authors,
            "sov": round(r.sov, 4), "velocity": r.velocity, "accel": r.accel,
            "z": (round(r.z, 3) if r.z is not None else None), "net_dir": round(r.net_dir, 3),
            "dd_count": r.dd_count, "baseline_status": r.baseline_status, "h_e": round(r.h_e, 4),
            "ret": (round(a.ret, 4) if a and a.ret is not None else None),
            "rvol": (round(a.rvol, 3) if a and a.rvol is not None else None),
            "rvol_conf": (a.rvol_conf if a else None),
            "h_m": (round(a.h_m, 4) if a else None),
        })
    payload = {
        "window_start": window_start,
        "window_end": window_start + window_seconds,
        "window_seconds": window_seconds,
        "generated_at": generated_at,
        "total_mentions": total_mentions,
        "quiet": (min_window_mentions is not None and total_mentions < min_window_mentions),
        # `capped`: the poll hit its pagination cap → the window is undercounted → `sov` is
        # untrustworthy (data-model invariant 14). Surfaced so the board can banner it low-trust.
        "capped": capped,
        "rows": out_rows,
    }
    if movers is not None:
        payload["movers"] = [
            {"symbol": m.symbol, "name": pretty_name(names.get(m.symbol)), "kind": m.kind,
             "rank": m.rank, "price": m.price, "percent_change": m.percent_change, "volume": m.volume}
            for m in movers
        ]
    # Atomic write: the dashboard reads this file concurrently; a torn read on a half-written
    # file would surface as transient "no data". Write to a temp sibling, then atomic replace.
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(payload, indent=2))
    tmp.replace(path)


def write_history(db: DB, path: Path) -> None:
    """Atomically refresh the history Parquet the dashboard's time views read (lock-free, like the
    JSON snapshot). Cheap full dump — the table is hourly-grained, so even years stay small."""
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    db.export_history_parquet(tmp)
    tmp.replace(path)


def run_aggregation(db: DB, settings, now: int, snapshot_path: Path | None = None,
                    names: dict | None = None, capped: bool = False):
    """Aggregate the window containing `now`, persist features, optionally write the snapshot.

    When a snapshot is written (i.e. for the current window, not the W−1 finalization pass) we also
    refresh history.parquet so the dashboard's day/month views stay current. `capped` (from the poll)
    flows into the snapshot so the board can banner an incomplete window (data-model invariant 14).
    """
    window_seconds = settings.ingest["window_seconds"]
    heat = settings.cfg["heat"]
    ws = window_start_for(now, window_seconds)
    rows = aggregate_window(
        db, ws, window_seconds,
        weights=heat["weights"],
        min_samples_ready=settings.cfg["baseline"]["min_samples_ready"],
        min_authors_full=heat.get("min_authors_full", 1.0),
    )
    db.upsert_empirical_features(rows)
    if snapshot_path is not None:
        if names is None:
            names = db.ticker_names()
        write_snapshot(rows, ws, window_seconds, snapshot_path, now, names=names,
                       min_window_mentions=heat.get("min_window_mentions"), capped=capped)
        write_history(db, settings.data_dir / "history.parquet")
    return ws, rows

File: packages/worker/src/aggregate.ts
/**
 * Windowed aggregator — TS port of the frozen v0.0.1 `wsb_signals/aggregate.py` (signal-framework
 * §2–§5). Turns the window's mentions into per-(ticker, window) empirical features + the composite
 * WSB Heat `H_e`. SoV-primary: `z` is computed but kept off the blend until its hour-of-week baseline
 * is `ready` (weight stays 0). Components are max-normalized (NOT percentile-ranked) within the window.
 *
 * Pure function: the Python original reads four things from DuckDB
 * (`mentions_in_window` / `features_at` / `sov_ranks_at` / `feature_history`); here those are explicit
 * inputs (captured in the golden fixtures), so the port is DB-free and diffs at the object boundary
 * (v2-porting-spec.md §1–§2). Gated on fixtures/aggregate/*.json — values AND order.
 *
 * Cross-language landmines handled:
 *  - `hourOfWeek` weekday remap: Python tm_wday (Mon=0) vs JS getUTCDay (Sun=0) → `((d+6)%7)*24+h`;
 *  - two-pass sample variance ÷ (n−1);
 *  - null `velocity`/`accel` propagation (no prior window vs. no prior velocity);
 *  - the blend is evaluated in the SAME operation order as Python so equal-input H_e is bit-identical,
 *    making the canonical tie-break (h_e→sov→authors→mentions→ticker) reproduce the oracle's order.
 *
 * NOTE (scope): `write_snapshot`/`pretty_name` are NOT ported here. The leaderboard.json snapshot is a
 * v0.0.1 presentation/locking workaround; in v2 the "publish" is a Postgres write + a freshness marker
 * (slice 3), and row rounding/pretty-names belong with the reader. `flairCounts` is an object (→ JSONB),
 * not Python's sorted json STRING — parity is on the COUNTS, not the serialization.
 */

/** A mention row as captured from `db.mentions_in_window`: [ticker, thingId, thingType, author, flair, direction]. */
export type MentionRow = readonly [
  ticker: string,
  thingId: string,
  thingType: string | null,
  author: string | null,
  flair: string | null,
  direction: string | null,
]

/** Prior-window features by ticker (from `features_at`): supplies mentions(W−1) + velocity(W−1). */
export type PriorFeatures = Record<string, { mentions: number | null; velocity: number | null }>

/** Per-(ticker, window, mentions) history rows (from `feature_history`): [ticker, windowStart, mentions]. */
export type HistoryRow = readonly [ticker: string, windowStart: number, mentions: number]

export interface HeatWeights {
  sov: number
  accel: number
  rank_delta: number
  authors: number
  conviction: number
  net_dir: number
  z: number
}

export interface AggregateInputs {
  windowStart: number
  windowSeconds: number
  weights: HeatWeights
  minSamplesReady: number
  minAuthorsFull: number
  mentionsInWindow: readonly MentionRow[]
  priorFeatures: PriorFeatures
  priorSovRanks: Record<string, number>
  featureHistory: readonly HistoryRow[]
}

export type BaselineStatus = 'cold' | 'warming' | 'ready'

/** A (ticker, window) empirical cell — the live empirical signal (mirrors `models.EmpiricalFeature`). */
export interface EmpiricalFeature {
  ticker: string
  windowStart: number
  mentions: number
  authors: number
  sov: number
  velocity: number | null
  accel: number | null
  z: number | null
  netDir: number
  ddCount: number
  flairCounts: Record<string, number> // {flair: count} — → JSONB (was a json string in DuckDB)
  baselineStatus: BaselineStatus
  hE: number
}

/** 0..167 — Monday 00:00 UTC = 0. JS `getUTCDay()` is Sun=0, so remap to Python's Mon=0 `tm_wday`. */
export function hourOfWeek(epoch: number): number {
  const d = new Date(epoch * 1000)
  const weekday = (d.getUTCDay() + 6) % 7 // Sun(0)→6, Mon(1)→0, … Sat(6)→5
  return weekday * 24 + d.getUTCHours()
}

/** Clock-aligned tumbling window containing `now`. */
export function windowStartFor(now: number, windowSeconds: number): number {
  return Math.floor(now / windowSeconds) * windowSeconds
}

/**
 * Scale to [0,1] by the window max (negatives floored to 0); all-zero or empty → zeros.
 * Max-norm — NOT percentile rank — keeps H_e SoV-primary (see aggregate.py `_max_norm` docstring).
 */
export function maxNorm(values: readonly number[]): number[] {
  const vmax = values.reduce((mx, v) => (v > mx ? v : mx), Number.NEGATIVE_INFINITY)
  if (values.length === 0 || vmax <= 0) return values.map(() => 0)
  return values.map((v) => Math.max(0, v) / vmax)
}

interface Feat {
  ticker: string
  mentions: number
  authors: number
  sov: number
  velocity: number | null
  accel: number | null
  netDir: number
  ddCount: number
  flairCounts: Record<string, number>
  z: number | null
  baselineStatus: BaselineStatus
  rankDelta: number
}

interface Bucket {
  things: Set<string>
  authors: Set<string>
  bull: number
  bear: number
  dd: Set<string>
  flairs: Map<string, number>
}

export function aggregateWindow(inp: AggregateInputs): EmpiricalFeature[] {
  const { windowStart, weights: w, minSamplesReady, minAuthorsFull } = inp
  if (inp.mentionsInWindow.length === 0) return []

  // group by ticker, dedup things by thingId (one mention per (ticker, thing) — §2.1 grain)
  const per = new Map<string, Bucket>()
  for (const [ticker, thingId, thingType, author, flair, direction] of inp.mentionsInWindow) {
    let d = per.get(ticker)
    if (!d) {
      d = { things: new Set(), authors: new Set(), bull: 0, bear: 0, dd: new Set(), flairs: new Map() }
      per.set(ticker, d)
    }
    if (d.things.has(thingId)) continue
    d.things.add(thingId)
    if (author) d.authors.add(author)
    if (direction === 'bull') d.bull++
    else if (direction === 'bear') d.bear++
    if (flair) {
      d.flairs.set(flair, (d.flairs.get(flair) ?? 0) + 1)
      if (thingType === 'post' && flair.trim().toUpperCase() === 'DD') d.dd.add(thingId)
    }
  }

  let total = 0
  for (const d of per.values()) total += d.things.size
  total = total || 1

  const priorExists = Object.keys(inp.priorFeatures).length > 0 // was W−1 aggregated at all?

  // baselines: prior windows in the same hour-of-week bucket (forward-only)
  const how = hourOfWeek(windowStart)
  const base = new Map<string, number[]>()
  for (const [tk, ws, m] of inp.featureHistory) {
    if (hourOfWeek(ws) === how) {
      const arr = base.get(tk)
      if (arr) arr.push(m)
      else base.set(tk, [m])
    }
  }

  // first pass — raw features per ticker
  const feats: Feat[] = []
  for (const [t, d] of per) {
    const m = d.things.size
    let velocity: number | null
    let accel: number | null
    if (priorExists) {
      // W−1 was aggregated, so a ticker absent from it genuinely had 0 mentions then.
      const pf = inp.priorFeatures[t]
      const pm = (pf?.mentions ?? 0) || 0 // Python `... or 0` — null/0 → 0
      velocity = m - pm
      const pv = pf?.velocity // accel needs W−1's stored velocity, else undefined
      accel = pv != null ? velocity - pv : null
    } else {
      // No prior window at all (cold start / gap): momentum is UNDEFINED — emitting 0-based deltas
      // would make every ticker look like a fresh breakout and inflate H_e.
      velocity = null
      accel = null
    }
    const bull = d.bull
    const bear = d.bear
    const netDir = bull + bear ? (bull - bear) / (bull + bear) : 0

    const samples = base.get(t) ?? []
    let status: BaselineStatus = 'cold'
    let z: number | null = null
    // max(2, …): sample variance divides by (n−1), so a misconfigured min_samples_ready=1 would
    // divide by zero — never treat a single sample as `ready`.
    if (samples.length >= Math.max(2, minSamplesReady)) {
      status = 'ready'
      const mean = samples.reduce((s, x) => s + x, 0) / samples.length
      // `(x-mean)*(x-mean)` not `**2`: explicit multiply is the correctly-rounded square in every engine,
      // bit-matching Python's `(x-mean) ** 2`; `Math.pow(_, 2)` is not guaranteed identical across engines.
      const variance = samples.reduce((s, x) => s + (x - mean) * (x - mean), 0) / (samples.length - 1)
      const sd = Math.sqrt(variance)
      z = sd > 0 ? (m - mean) / sd : null
    } else if (samples.length > 0) {
      status = 'warming'
    }

    feats.push({
      ticker: t,
      mentions: m,
      authors: d.authors.size,
      sov: m / total,
      velocity,
      accel,
      netDir,
      ddCount: d.dd.size,
      // sorted keys — matches Python `json.dumps(..., sort_keys=True)`; harmless for JSONB equality but
      // keeps a canonical key order if anything later stringifies it for hashing/diffing.
      flairCounts: Object.fromEntries([...d.flairs].sort((a, b) => cmpStr(a[0], b[0]))),
      z,
      baselineStatus: status,
      rankDelta: 0,
    })
  }

  // rank_delta: prior rank − current rank (+ve = climbing). Tie-break equal-SoV by ticker so rank is
  // deterministic, not dependent on iteration order.
  const ranked = [...feats].sort((a, b) => b.sov - a.sov || cmpStr(a.ticker, b.ticker))
  const curRank = new Map<string, number>()
  ranked.forEach((f, i) => curRank.set(f.ticker, i + 1))
  for (const f of feats) {
    const pr = inp.priorSovRanks[f.ticker]
    f.rankDelta = pr ? pr - curRank.get(f.ticker)! : 0 // Python `if pr` — ranks are ≥1; absent → 0
  }

  // normalize components to [0,1] by window max, then weighted blend → H_e (SoV-primary)
  const sovN = maxNorm(feats.map((f) => f.sov))
  const accN = maxNorm(feats.map((f) => (f.accel != null ? f.accel : 0))) // None/neg → 0
  const rdN = maxNorm(feats.map((f) => f.rankDelta))
  const auN = maxNorm(feats.map((f) => f.authors))
  const ddN = maxNorm(feats.map((f) => f.ddCount))
  const zN = maxNorm(feats.map((f) => (f.z != null ? f.z : 0)))

  const out: EmpiricalFeature[] = feats.map((f, i) => {
    const zReady = f.baselineStatus === 'ready' && f.z != null
    let hE =
      w.sov * sovN[i]! +
      w.accel * accN[i]! +
      w.rank_delta * rdN[i]! +
      w.authors * auN[i]! +
      w.conviction * ddN[i]! +
      w.net_dir * Math.abs(f.netDir) +
      (zReady ? w.z * zN[i]! : 0)
    // Support shrink: damp absolute H_e toward 0 until `minAuthorsFull` distinct authors back the row,
    // so a lone mention (every max-normed component = 1.0 in a quiet window) can't top the board.
    if (minAuthorsFull > 0) hE *= Math.min(1, f.authors / minAuthorsFull)
    return {
      ticker: f.ticker,
      windowStart,
      mentions: f.mentions,
      authors: f.authors,
      sov: f.sov,
      velocity: f.velocity,
      accel: f.accel,
      z: f.z,
      netDir: f.netDir,
      ddCount: f.ddCount,
      flairCounts: f.flairCounts,
      baselineStatus: f.baselineStatus,
      hE,
    }
  })

  // Canonical board order: H_e-primary, explicit total-order tie-break down to `ticker` so equal-H_e
  // rows have ONE deterministic ordering (must not depend on iteration order). h_e→sov→authors→mentions→ticker.
  //
  // RAW float compare — NOT quantized. The port is bit-exact to the oracle (same IEEE ops, same order),
  // so equal rows compare equal and genuinely-different rows compare like Python's raw sort — including
  // when the oracle's order rests on a 1-ULP h_e difference (e.g. 0.4 vs 0.39999999999999997). Quantizing
  // to 1e-9 (porting-spec §2.6, original) DISCARDS that real signal and inverts such pairs vs the oracle —
  // the random/010 fixture proved it. Near-tie robustness belongs in the slice-9 shadow-diff tolerance,
  // not in the production sort. (Spec §2.6 corrected to match.)
  out.sort(compareBoard)
  return out
}

/** A row carrying the canonical-sort keys — the structural shape `compareBoard` orders on. */
export type BoardRow = Pick<EmpiricalFeature, 'hE' | 'sov' | 'authors' | 'mentions' | 'ticker'>

/**
 * The canonical leaderboard total order — `h_e→sov→authors→mentions→ticker`, on RAW floats (see the
 * call-site note above). Exported so the rank/rank_delta reads (db.readHeRanksAt) order identically to
 * `aggregateWindow`'s board — one definition, no drift between the live board and the persisted ranks.
 */
export function compareBoard(a: BoardRow, b: BoardRow): number {
  return b.hE - a.hE || b.sov - a.sov || b.authors - a.authors || b.mentions - a.mentions || cmpStr(a.ticker, b.ticker)
}

/** Ascending string compare matching Python's `<` on str (code-point order for the ASCII tickers here). */
function cmpStr(a: string, b: string): number {
  return a < b ? -1 : a > b ? 1 : 0
}
